pkg/shell/cockpit-server.js

```diff
@@ -824,54 +849,120 @@ PageShutdownDialog.prototype = {
         }
 
-        $("#shutdown-action").prop('disabled', disabled);
+        this.show_errors(! valid);
+        return valid;
+    },
+
+    update: function() {
+        var delay = shell.select_btn_selected(this.delay_btn);
+        $("#shutdown-time").toggle(delay == "x");
+
+        this.show_errors(false);
     },
 
     do_action: function(op) {
+        var self = this;
+        var dfd = $.Deferred();
         var delay = shell.select_btn_selected(this.delay_btn);
-        var message = $("#shutdown-message").val();
-        var when;
 
-        if (delay == "x")
-            when = ($("#shutdown-time input:nth-child(1)").val() + ":" +
-                    $("#shutdown-time input:nth-child(3)").val());
-        else
-            when = "+" + delay;
+        function get_server_date() {
+            if (delay == "x") {
+                var datestr = $("#shutdown-date-input").val();
+                var hourstr = $("#shutdown-time-hours").val();
+                var minstr = $("#shutdown-time-minutes").val();
 
-        var arg = (op == "shutdown") ? "--poweroff" : "--reboot";
-        cockpit.spawn(["shutdown", arg, when, message], { superuser: "try" })
-            .fail(function(ex) {
-                $('#shutdown-dialog').modal('hide');
-                shell.show_unexpected_error(ex);
-            })
-            .done(function(ex) {
-                if (op == "restart")
-                    cockpit.hint("restart");
-                $('#shutdown-dialog').modal('hide');
-            });
+                cockpit.spawn(["/usr/bin/date", "--date=" + datestr + " " + hourstr + ":" + minstr, "+%s"])
+                    .fail(function(ex) {
+                        dfd.reject(ex);
+                    })
+                    .done(function(data) {
+                        var input_timestamp = parseInt(data, 10);
+                        var server_timestamp = parseInt(PageShutdownDialog.server_time.raw_now.getTime() / 1000, 10);
+                        var when = Math.ceil((input_timestamp - server_timestamp) / 60);
+
+                        dfd.resolve(when);
+                    });
+            } else {
+                dfd.resolve(delay);
+            }
+
+            return dfd.promise();
+        }
+
+        if (! this.validate())
+            return;
+
+        $.when(get_server_date()).done(function(when) {
+
+            if (when < 0) {
+                $('#shutdown-parse-error').text(_("Cannot schedule event in the past"));
+                self.show_errors(true);
+            } else {
+                when = "+" + when;
+            }
+
+            var arg = (op == "shutdown") ? "--poweroff" : "--reboot";
+            var message = $("#shutdown-message").val();
+            cockpit.spawn(["shutdown", arg, when, message], { superuser: true })
+                .fail(function(ex) {
+                    $('#shutdown-dialog').modal('hide');
+                    shell.show_unexpected_error(ex);
```

Comment:
This should be shown in the dialog, as described here: https://github.com/cockpit-project/cockpit/wiki/HIG:-Dialogs

Here's an example of showing a general failure in a dialog. https://github.com/cockpit-project/cockpit/blob/master/pkg/playground/patterns.html#L85

In addition the dialog should not close on failure.